cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ETLStage.java

```diff
@@ -77,6 +104,117 @@ public ETLStage upgradeStage(UpgradeContext upgradeContext) {
+      new ArtifactSelectorConfig(newPlugin.getScope().name(), newPlugin.getName(),
+                                 newVersion);
+    io.cdap.cdap.etl.proto.v2.ETLPlugin upgradedEtlPlugin =
+      new io.cdap.cdap.etl.proto.v2.ETLPlugin(plugin.getName(), plugin.getType(),
+                                              plugin.getProperties(),
+                                              newArtifactSelectorConfig);
+    return upgradedEtlPlugin;
+  }
+
+  /**
+   * Returns new valid version string for plugin upgrade if any changes are required. Returns null if no change to
+   * current plugin version.
+   * Artifact selector config only stores plugin version as string, it can be either fixed version or range.
+   * Hence, if the plugin version is fixed, replace the fixed version with newer fixed version. If it is a range,
+   * move the upper bound of the range to the newest version.
+   *
+   * @param newPlugin New candidate plugin for updating plugin artifact.
+   * @return version string to be used for new plugin. Might be fixed version/version range string depending on
+   *         current use.
+   */
+  private String getUpgradedVersionString(ArtifactId newPlugin) {
+    try {
+      ArtifactVersionRange currentVersionRange =
+        io.cdap.cdap.api.artifact.ArtifactVersionRange.parse(plugin.getArtifactConfig().getVersion());
+      if (!currentVersionRange.isExactVersion()) {
+        // Current plugin version is version range.
+        if (currentVersionRange.versionIsInRange(newPlugin.getVersion())) {
+          // Do nothing and return as is. Note that plugin scope will not change.
+          // TODO: Figure out how to change plugin scope if a newer plugin is found but in different scope.
+          return null;
+        } else if (currentVersionRange.getLower().compareTo(newPlugin.getVersion()) > 0) {
+          // Current lower version is higher than newer latest version. This should not happen.
+          LOG.warn("Invalid new plugin artifact {} for upgrade for upgrading plugin {} ", newPlugin, plugin);
+          return null;
+        } else {
+          // Increase the upper bound to latest available version.
+          ArtifactVersionRange newVersionRange =
+            new ArtifactVersionRange(currentVersionRange.getLower(), currentVersionRange.isLowerInclusive(),
+                                     newPlugin.getVersion(), /*isUpperInclusive=*/true);
+          return newVersionRange.getVersionString();
+        }
+      } else if (currentVersionRange.isExactVersion()
+          && currentVersionRange.getLower().compareTo(newPlugin.getVersion()) < 0) {
+        // Current version is a fixed version and new version is higher than current.
+        return newPlugin.getVersion().getVersion();
+      } else {
+        // Should not happen. Safety check.
+        return null;
+      }
+    } catch (Exception e) {
+      LOG.warn("Issue in parsing version string for plugin, ignoring stage for upgrade {}", plugin, e);
```

Comment:
there are too many levels of nesting in this method. If the parsing is the only thing that can throw an exception, you remove one level of nesting by doing:

```
ArtifactVersionRange currentVersionRange;
try {
  currentVersionRange = ...
} catch (...) {
  ...
}

// rest of logic
```

Reply:
I agree. I simplified it further a bit.